designs/odcr.md

```diff
@@ -0,0 +1,204 @@
+By default it will not fallback to `on-demand` instances, if not specified specifically in the capacity-type in the NodePool as described above [capacity-reservation-fleet (falling back to on-demand if capacity-reservation not available)](#capacity-reservation-fleet-falling-back-to-on-demand-if-capacity-reservation-not-available).
+
+To clarify the NodeClass with Capacity Reservations will pin the NodeClass into all instance types and availability zones the Capacity Reservations reserved, if additionally on-demand is provided it can spin up other instance types, but can have unintended side effects during consolidation phase.
+
+_Note that these aren't permanent restrictions but simply narrowing down what features exist in the first iteration of supporting Capacity Reservation_ 
+
+Pros: 
+- This makes the user responsible for verifying the capability of their NodeClasses and requirements. It makes no assumption about what a Capacity Reservation can or cannot do
+- Users are still open to define fallbacks of on-demand or spot, and its up them of how they would like to configure it. Making it possible to either define everything with one EC2NodeClass and one NodePool, or using multiple with weights
+
+Cons:
+- This implementation is overly restrictive and will be difficult to use when attempting to utilize multiple Instance Types. Because ODCRs are such a specific use case, it can be argued that the flexibility of multiple Instance Types can cause issues with cost savings
+
+#### Pricing and consolidation
+
+##### Provisioning
+
+Pricing is directly considered during provisioning and consolidation as capacity-reservation is prepaid. It is assumed to have a price of 0 during provisioning.
+
+##### Consolidating Capacity Reserved Instances
+
+During consolidation pricing does matter as it affects which candidate will be [prioritized](https://github.com/kubernetes-sigs/karpenter/blob/75826eb51589e546fffb594bfefa91f3850e6c82/pkg/controllers/disruption/consolidation.go#L156). Since all capacity instances are paid ahead of time, their cost is already incurred. Users would likely want to prioritize filling their reserved capacity.
+reservation first then fall back into other instances. Because of this reserved instances should likely show up as 0 dollar pricing when we calculate the instance pricing. Since each candidate is tied to a NodePool and EC2NodeClass, we should be able to safely override the pricing per node under a capacity reservation.
```

Comment:
What about if the ODCR gets canceled? The instance will still be running but no longer "0 dollars" - will Karpenter poll to account for this?

Reply:
sure, we can add all the properties from the struct https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/service/ec2@v1.162.1/types#CapacityReservation

Reply:
Discussed offline: Karpenter will be polling instances that it owns and validating that they are still in an ODCR. If they are no longer in an ODCR, it will remove the ODCR label the NodeClaim and Node to reflect this. This ensures that new pods that would select on ODCR will not schedule to these nodes and also ensures that we detect compatibility changes that would cause us to drift when a user has a NodePool that specifically targets ODCRs but has a Node in the NodePool that no longer exists in any ODCR